backend/server.js

```diff
@@ -213,6 +216,111 @@ sbp('sbp/selectors/register', {
+  },
+  async 'backend/deleteFile' (cid: string): Promise<void> {
+    const owner = await sbp('chelonia.db/get', `_private_owner_${cid}`)
+    const rawManifest = await sbp('chelonia.db/get', cid)
+    if (rawManifest === '') throw new BackendErrorGone()
+    if (!rawManifest) throw new BackendErrorNotFound()
+
+    try {
+      const manifest = JSON.parse(rawManifest)
+      if (!manifest || typeof manifest !== 'object') throw new BackendErrorBadData('manifest format is invalid')
+      if (manifest.version !== '1.0.0') throw new BackendErrorBadData('unsupported manifest version')
+      if (!Array.isArray(manifest.chunks) || !manifest.chunks.length) throw BackendErrorBadData('missing chunks')
+      // Delete all chunks
+      await Promise.all(manifest.chunks.map(([, cid]) => sbp('chelonia.db/delete', cid)))
+    } catch (e) {
+      console.warn(e, `Error parsing manifest for ${cid}. It's probably not a file manifest.`)
+      throw new BackendErrorNotFound()
+    }
+    // The keys to be deleted are not read from or updated, so they can be deleted
+    // without using a queue
+    const resourcesKey = `_private_resources_${owner}`
+    await removeFromIndexFactory(resourcesKey)(cid)
+
+    await sbp('chelonia.db/delete', `_private_owner_${cid}`)
+    await sbp('chelonia.db/delete', `_private_size_${cid}`)
+    await sbp('chelonia.db/delete', `_private_deletionToken_${cid}`)
+
+    await sbp('chelonia.db/set', cid, '')
+  },
+  // eslint-disable-next-line require-await
+  async 'backend/deleteContract' (cid: string): Promise<void> {
+    let contractsPendingDeletion = sbp('okTurtles.data/get', 'contractsPendingDeletion')
+    if (!contractsPendingDeletion) {
+      contractsPendingDeletion = new Set()
+      sbp('okTurtles.data/set', 'contractsPendingDeletion', contractsPendingDeletion)
+    }
+    // Avoid deadlocks due to loops
+    if (contractsPendingDeletion.has(cid)) {
+      return
+    }
+    contractsPendingDeletion.add(cid)
+
+    return sbp('chelonia/queueInvocation', cid, async () => {
+      const owner = await sbp('chelonia.db/get', `_private_owner_${cid}`)
+      const rawManifest = await sbp('chelonia.db/get', cid)
+      if (rawManifest === '') throw new BackendErrorGone()
+      if (!rawManifest) throw new BackendErrorNotFound()
+
+      const resourcesKey = `_private_resources_${cid}`
+      const resources = await sbp('chelonia.db/get', resourcesKey)
+      if (resources) {
+        await Promise.allSettled(resources.split('\x00').map((resourceCid) => {
+          const parsed = parseCID(resourceCid)
+
+          if (parsed.code === multicodes.SHELTER_CONTRACT_DATA) {
+            return sbp('chelonia.persistentActions/enqueue', ['backend/deleteContract', resourceCid])
+          } else if (parsed.code === multicodes.SHELTER_FILE_MANIFEST) {
+            return sbp('chelonia.persistentActions/enqueue', ['backend/deleteFile', resourceCid])
+          }
+
+          return undefined
+        }))
+      }
```

Comment:
Could you add some comments here explaining what this block is doing?

I assume it's deleting resources related to this contract, e.g. file attachments in a chatroom, and associated contracts (e.g. group => chatroom contracts)